### PR TITLE
Add favouring shops

### PR DESCRIPTION
**Pull request by Almond-Peach**
[As a user I should be able to mark the shops as favorite](https://trello.com/c/pO1e2iqm)

### CODE REVIEW

**Comment by Almond-Peach** [comment]
This Pull Request should be merged after [Add shop details screen #21](https://github.com/SimplyTechnologies/android-internship/pull/21)

File: app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt [hunk 1]
```diff
@@ -1,27 +1,51 @@
...
-import com.simply.birthdayapp.presentation.ui.models.Shop
+import com.simply.birthdayapp.presentation.models.Shop
 import com.simply.birthdayapp.type.ShopFilter
 import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
 
 interface ShopsRepository {
```
Comment (ManukyanQristine) [inline app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt:13]: For failure case please use error messages coming from response.

File: app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt [hunk 2]
```diff
@@ -1,27 +1,51 @@
...
+    override fun addShopToFavourites(shopId: Int): Flow<Result<Int>> = flow {
+        val response = apolloClient.mutation(AddShopToFavouriteMutation(shopId)).execute()
+        if (response.hasErrors()) {
+            emit(Result.failure(Throwable()))
+        } else {
+            val responseShopId = response.data?.addShopToFavorite?.shopId
+            if (responseShopId == null) emit(Result.failure(Throwable()))
+            else emit(Result.success(responseShopId))
+        }
+    }.flowOn(Dispatchers.IO)
```
Comment (Hayk985) [inline app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt:39]: Aren't you already inside the IO dispatcher? Do we need to specify it here too?

Comment (Hayk985) [inline app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt]: Aren't you already inside the IO dispatcher? Do we need to specify it here too?

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/components/ShopCard.kt [hunk 3]
```diff
@@ -48,13 +53,32 @@ fun ShopCard(shop: Shop) {
...
+                    if (shop.isFavourite) AppTheme.colors.gray
+                    else AppTheme.colors.lightPink,
+                )
+            } else {
+                IconButton(
+                    modifier = Modifier.size(24.dp),
+                    onClick = { onIsFavouriteChange(shop) },
+                ) {
+                    Icon(
+                        painter = if (shop.isFavourite) painterResource(id = R.drawable.ic_heart_filled)
```
Comment (Hayk985) [inline app/src/main/java/com/simply/birthdayapp/presentation/ui/components/ShopCard.kt]: you can write the condition inside `painterResource`

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsScreen.kt [hunk 4]
```diff
@@ -27,31 +36,81 @@ import com.simply.birthdayapp.R
...
 
     LaunchedEffect(shopsLazyListState) {
         snapshotFlow { shopsLazyListState.firstVisibleItemIndex }
             .debounce(500L)
             .collectLatest { shopsViewModel.onScrollPositionChange(it) }
     }
 
-    val focusManager = LocalFocusManager.current
+    LaunchedEffect(lastFavouredShopName) {
+        val shopName = lastFavouredShopName
```
Comment (Hayk985) [inline app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsScreen.kt]: Wrap lastFavouredShopName with let

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt [hunk 5]
```diff
@@ -52,29 +62,112 @@ class ShopsViewModel(
         _searchBarQuery.update { searchBarQuery }
     }
 
+    fun onShopIsFavouriteChange(shop: Shop) {
+        viewModelScope.launch(Dispatchers.IO) {
+            val cachedShopIndex = _cachedShops.indexOfFirst { it.id == shop.id }.takeIf { it != -1 }
```
Comment (Hayk985) [inline app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt]: Handle null case and remove make argument not-nullable.

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt [hunk 6]
```diff
@@ -52,29 +62,112 @@ class ShopsViewModel(
...
             }
+            .catch { _lastShopsError.update { ShopsError.RemoveShopFromFavourites } }
+            .flowOn(Dispatchers.IO)
+            .launchIn(viewModelScope)
+    }
+
+    private suspend fun setCachedShopIsLoadingFavourite(
+        cachedShopIndex: Int?,
+        isLoadingFavourite: Boolean,
+    ) = withContext(Dispatchers.IO) {
```
Comment (Hayk985) [inline app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt]: Use Dispatchers.Default

Comment (Almond-Peach) [reply app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt]: Fixed

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt (hunk 5 above)
Comment (Almond-Peach) [reply app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsViewModel.kt]: Fixed

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsScreen.kt (hunk 4 above)
Comment (Almond-Peach) [reply app/src/main/java/com/simply/birthdayapp/presentation/ui/screens/main/shops/ShopsScreen.kt]: Fixed

File: app/src/main/java/com/simply/birthdayapp/presentation/ui/components/ShopCard.kt (hunk 3 above)
Comment (Almond-Peach) [reply app/src/main/java/com/simply/birthdayapp/presentation/ui/components/ShopCard.kt]: Fixed

File: app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt (hunk 2 above)
Comment (Almond-Peach) [reply app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt]: Fixed

File: app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt (hunk 1 above)
Comment (Almond-Peach) [reply app/src/main/java/com/simply/birthdayapp/data/repositories/ShopsRepository.kt:13]: Updated

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -41,4 +49,15 @@
     <string name="ok">OK</string>
     <string name="registration_error">Registration Error</string>
     <string name="network_error">Sorry, an error occurred registering your account, please try again.</string>
+    <string name="shop_website">Shop website</string>
+    <string name="shop_phone_number">"Phone: "</string>
+    <string name="shop_phone_number_not_specified">Phone: not specified</string>
+    <string name="shop_address">"Address: "</string>
+    <string name="google_maps_search_url"><![CDATA[https://www.google.com/maps/search/?api=1&query=%1$s]]></string>
```
Comment (ManukyanQristine) [inline app/src/main/res/values/strings.xml]: Please move  google_maps_search_url to local.properties

Comment (Almond-Peach) [reply app/src/main/res/values/strings.xml]: Moved

**Review by ManukyanQristine — APPROVED** [review]

**Pull request merged**